src/mx_bluesky/common/preprocessors/preprocessors.py

```diff
@@ -0,0 +1,108 @@
+
+    XBPM feedback isn't reliable during collections due to:
+     * Objects (e.g. attenuator) crossing the beam can cause large (incorrect) feedback movements
+     * Lower transmissions/higher energies are less reliable for the xbpm
+
+    So we need to keep the transmission at 100% and the feedback on when not collecting
+    and then turn it off and set the correct transmission for collection. The feedback
+    mostly accounts for slow thermal drift so it is safe to assume that the beam is
+    stable during a collection.
+
+    In the case of a beam dump, undulator gap may not return. Therefore, we check here
+    that the undulator gap is correct after XBPM is stable, and before collection.
+
+    Args:
+        plan: The plan performing the data collection.
+        devices (DevicesForXbpmAndTransmissionWrapper): Composite device including The XBPM device that is responsible for keeping
+                                                        the beam in position, undulator, attenuator, and DCM
+        desired_transmission_fraction (float): The desired transmission for the collection
+        run_key_to_wrap: (str | None): Pausing XBPM and setting transmission is inserted after the 'open_run' message is seen with the matching run key, and unpausing and resetting transmission is inserted after the corresponding 'close_run' message is seen. If not specified, instead wrap the first run encountered.
+    """
+
+    _wrapped_plan_name: None | str = None
+
+    def head(msg):
+        yield from check_and_pause_feedback_and_verify_undulator_gap(
+            devices.undulator,
+            devices.xbpm_feedback,
+            devices.attenuator,
+            devices.dcm,
+            desired_transmission_fraction,
+        )
+
+        # Allow 'open_run' message to pass through
+        yield msg
+
+    def tail():
+        yield from unpause_xbpm_feedback_and_set_transmission_to_1(
+            devices.xbpm_feedback, devices.attenuator
+        )
+
+    def insert_plans(msg):
+        # Wrap the specified run, or, if none specified, wrap the first run encountered
+        nonlocal _wrapped_plan_name
+
+        if msg.command:
+            # If we specified a run key, did we encounter it
+            # If we didn't specify, then insert the plans and track the name of the run
+            if msg.command == "open_run" and (
+                run_key_to_wrap
+                and run_key_to_wrap is msg.run
+                or (not (run_key_to_wrap or _wrapped_plan_name))
+            ):
+                _wrapped_plan_name = msg.run if msg.run else "unnamed_run"
+
+                return head(msg), None
+
+            # Check if the run tracked from above was closed
+            elif msg.command == "close_run" and (
+                _wrapped_plan_name == "unnamed_run"
+                or (msg.run and _wrapped_plan_name and _wrapped_plan_name is msg.run)
+            ):
+                return None, tail()
```

Comment:
Could: This may be a bit cleaner using pattern matching:

```suggestion
        match msg.command:
            case "open_run":
                # If we specified a run key, did we encounter it
                # If we didn't specify, then insert the plans and track the name of the run
                if (
                    not (run_key_to_wrap or _wrapped_plan_name)
                    or run_key_to_wrap is msg.run
                ):
                    _wrapped_plan_name = msg.run if msg.run else "unnamed_run"

                    return head(msg), None
            case "close_run":
                # Check if the run tracked from above was closed
                if _wrapped_plan_name == "unnamed_run" or (
                    msg.run and _wrapped_plan_name and _wrapped_plan_name is msg.run
                ):
                    return None, tail()
```